spec/light-client/Lightclient.tla

```diff
@@ -1,111 +1,273 @@
   MAX_POWER
-    (* The maximal voting power of a single node *)  
-    
-Heights == 0..ULTIMATE_HEIGHT
-
-AllNodes == Corr \union Faulty
-
-Powers == 1..MAX_POWER
-
-(* A commit is just a set of nodes who have committed the block *)
-Commits == SUBSET AllNodes
-
-(* The set of all block headers that can be on the blockchain.
-   This is a simplified version of the Block data structure in the actual implementation. *)
-BlockHeaders == [
-  height: Heights,
-    \* the block height
-  lastCommit: Commits,
-    \* the nodes who have voted on the previous block, the set itself instead of a hash
-  (* in the implementation, only the hashes of V and NextV are stored in a block,
-     as V and NextV are stored in the application state *) 
-  V: [AllNodes -> Powers],
-    \* the validators of this block together with their voting powers,
-    \* i.e., a multi-set. We store the validators instead of the hash.
-  NextV: [AllNodes -> Powers]
-    \* the validators of the next block together with their voting powers,
-    \* i.e., a multi-set. We store the next validators instead of the hash.
-]
-
-SignedHeaders == BlockHeaders \X Commits
+    (* a maximal voting power of a single node *)
 
-VARIABLES
-    height,
-    (* the height of the blockchain, starting with 0 *)
-    minTrustedHeight,
-    (* The global height of the oldest block that is younger than
-       the trusted period (AKA the almost rotten block).
-       In the implementation, this is the oldest block,
-       where block.bftTime + trustingPeriod >= globalClock.now. *)
-    blockchain
-    (* A sequence of BlockHeaders,
-       which gives us the God's (or Daemon's) view of the blockchain. *)
-
-(* Compute the total voting power of a subset of validators Nodes,
-   whose individual voting powers are given by a function vp *)  
-RECURSIVE PowerOfSet(_, _)
-PowerOfSet(vp, Nodes) ==
-    IF Nodes = {}
-    THEN 0
-    ELSE LET node == CHOOSE n \in Nodes: TRUE IN
-        (* ASSERT(node \in DOMAIN vp) *)
-        vp[node] + PowerOfSet(vp, Nodes \ {node})
-
-(* Is the voting power correct,
-   that is, more than 2/3 of the voting power belongs to the correct processes? *)
-IsCorrectPower(vp) ==
-    LET CV == Corr \intersect DOMAIN vp
-        FV == Faulty \intersect DOMAIN vp
-        CP == PowerOfSet(vp, CV)
-        FP == PowerOfSet(vp, FV)
-    IN
-    CP > 0 /\ CP > 2 * FP
+\* the state variables of Blockchain, see Blockchain.tla for the details
+VARIABLES tooManyFaults, height, minTrustedHeight, blockchain, Faulty
+
+\* All the variables of Blockchain. For some reason, BC!vars does not work
+bcvars == <<tooManyFaults, height, minTrustedHeight, blockchain, Faulty>>
+
+(* Create an instance of Blockchain.
+   We could write EXTENDS Blockchain, but then all the constants and state variables
+   would be hidden inside the Blockchain module.
+ *) 
+BC == INSTANCE Blockchain WITH tooManyFaults <- tooManyFaults, height <- height,
```

Comment:
Why we need to do this?

Reply:
An alternative would be to write `EXTENDS Blockchain` to include in the spec. But then Blockchain's parameters would be implicit, which I do not like.